tests/cpp_test/test_ChunkedArray.cpp

```diff
@@ -0,0 +1,262 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *
+ * Author: Alberto Ferreira
+ */
+#include <gtest/gtest.h>
+#include "../include/LightGBM/utils/ChunkedArray.hpp"
+
+using LightGBM::ChunkedArray;
+
+/*!
+  Helper util to compare two vectors.
+
+  Don't compare floating point vectors this way!
+*/
+template <typename T>
+testing::AssertionResult are_vectors_equal(const std::vector<T> &a, const std::vector<T> &b) {
+  if (a.size() != b.size()) {
+    return testing::AssertionFailure()
+      << "Vectors differ in size: "
+      << a.size() << " != " << b.size();
+  }
+
+  for (size_t i = 0; i < a.size(); ++i) {
+    if (a[i] != b[i]) {
+      return testing::AssertionFailure()
+        << "Vectors differ at least at position " << i << ": "
+        << a[i] << " != " << b[i];
+    }
+  }
+
+  return testing::AssertionSuccess();
+}
+
+
+class ChunkedArrayTest : public testing::Test {
+ protected:
+
+  void SetUp() override {
+
+  }
+
+  void add_items_to_array(const std::vector<int> &vec, ChunkedArray<int> &ca) {
+    for (auto v: vec) {
+      ca.add(v);
+    }
+  }
+
+  /*!
+    Ensures that if coalesce_to() is called upon the ChunkedArray,
+    it would yield the same contents as vec
+  */
+  testing::AssertionResult coalesced_output_equals_vec(const ChunkedArray<int> &ca, const std::vector<int> &vec,
+                                                       const bool all_addresses=false) {
+    std::vector<int> out(vec.size());
+    ca.coalesce_to(out.data(), all_addresses);
+    return are_vectors_equal(out, vec);
+  }
+
+  // Constants
+  const std::vector<int> REF_VEC = {1, 5, 2, 4, 9, 8, 7};
+  const size_t CHUNK_SIZE = 3;
+  const size_t OUT_OF_BOUNDS_OFFSET = 4;
+
+  ChunkedArray<int> ca_ = ChunkedArray<int>(CHUNK_SIZE); //<! Re-used for many tests.
+};
+
+
+/*! ChunkedArray cannot be built from chunks of size 0. */
+TEST_F(ChunkedArrayTest, constructorWithChunkSize0Throws) {
+  ASSERT_THROW(ChunkedArray<int> ca(0), std::length_error);
```

Comment:
If you are OK to use `Log::Fatal` for the consistency in `chunked_array.hpp`, `Log::Fatal` throws `runtime_error`:
https://github.com/microsoft/LightGBM/blob/4580393f604d825c318c053891f2870e6a40347f/include/LightGBM/utils/log.h#L129
```suggestion
  ASSERT_THROW(ChunkedArray<int> ca(0), std::runtime_error);
```